File: synap-server/src/core/geospatial.rs
//! Geospatial data structure implementation for Synap
//!
//! Provides Redis-compatible geospatial operations (GEOADD, GEODIST, GEORADIUS, etc.)
//! Storage: Backed by Sorted Sets with geohash-encoded scores
//!
//! # Architecture
//! ```text
//! GeospatialStore
//!   ├─ Uses SortedSetStore internally
//!   ├─ Converts lat/lon to geohash score
//!   └─ Provides geospatial query operations
//! ```
//!
//! # How Redis Stores Geodata
//! Redis uses Sorted Sets where:
//! - Member = location name (string)
//! - Score = geohash encoded as 52-bit integer
//!
//! The geohash is encoded as: `(lat + 90.0) * (1 << 26) + (lon + 180.0)`

use super::error::{Result, SynapError};
use super::sorted_set::SortedSetStore;
use geohash::{Coord, encode};
use parking_lot::RwLock;
use serde::{Deserialize, Serialize};
use std::sync::Arc;

const EARTH_RADIUS_KM: f64 = 6371.0;
const EARTH_RADIUS_M: f64 = 6371000.0;
const EARTH_RADIUS_MI: f64 = 3959.0;
const EARTH_RADIUS_FT: f64 = 20902231.0;

/// Distance unit for GEODIST
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum DistanceUnit {
    Meters,
    Kilometers,
    Miles,
    Feet,
}

impl DistanceUnit {
    /// Get Earth radius for this unit
    fn radius(&self) -> f64 {
        match self {
            DistanceUnit::Meters => EARTH_RADIUS_M,
            DistanceUnit::Kilometers => EARTH_RADIUS_KM,
            DistanceUnit::Miles => EARTH_RADIUS_MI,
            DistanceUnit::Feet => EARTH_RADIUS_FT,
        }
    }
}

impl std::str::FromStr for DistanceUnit {
    type Err = SynapError;

    fn from_str(s: &str) -> std::result::Result<Self, Self::Err> {
        match s.to_uppercase().as_str() {
            "M" | "METER" | "METERS" => Ok(DistanceUnit::Meters),
            "KM" | "KILOMETER" | "KILOMETERS" => Ok(DistanceUnit::Kilometers),
            "MI" | "MILE" | "MILES" => Ok(DistanceUnit::Miles),
            "FT" | "FOOT" | "FEET" => Ok(DistanceUnit::Feet),
            _ => Err(SynapError::InvalidValue(format!(
                "Invalid distance unit: {}. Must be one of: m, km, mi, ft",
                s
            ))),
        }
    }
}

/// Coordinate (latitude, longitude)
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct Coordinate {
    pub lat: f64,
    pub lon: f64,
}

impl Coordinate {
    pub fn new(lat: f64, lon: f64) -> Result<Self> {
        if !(-90.0..=90.0).contains(&lat) {
            return Err(SynapError::InvalidValue(format!(
                "Latitude must be between -90 and 90, got: {}",
                lat
            )));
        }
        if !(-180.0..=180.0).contains(&lon) {
            return Err(SynapError::InvalidValue(format!(
                "Longitude must be between -180 and 180, got: {}",
                lon
            )));
        }
        Ok(Self { lat, lon })
    }
}

/// Convert coordinate to geohash score (Redis-compatible encoding)
/// Redis uses: (lat + 90.0) * (1 << 26) + (lon + 180.0)
/// This produces a 52-bit integer stored as f64
fn coordinate_to_score(lat: f64, lon: f64) -> f64 {
    // Normalize to [0, 180) for lat and [0, 360) for lon
    let normalized_lat = (lat + 90.0).clamp(0.0, 180.0);
    let normalized_lon = (lon + 180.0).clamp(0.0, 360.0);

    // Encode as 52-bit integer (26 bits for lat, 26 bits for lon)
    // Normalize to integer range: [0, 180*2^26) for lat, [0, 360*2^26) for lon
    // But we need to fit lon in 26 bits, so scale it to [0, 2^26)
    let lat_scaled = (normalized_lat * (1u64 << 26) as f64 / 180.0).round() as u64;
    let lon_scaled = (normalized_lon * (1u64 << 26) as f64 / 360.0).round() as u64;

    // Combine: lat in upper 26 bits, lon in lower 26 bits
    ((lat_scaled << 26) | lon_scaled) as f64
}

/// Convert geohash score back to coordinate
fn score_to_coordinate(score: f64) -> Coordinate {
    // Extract lat and lon from encoded score
    // Score format: (lat_scaled << 26) | lon_scaled
    let bits = score as u64;
    let lat_scaled = bits >> 26;
    let lon_scaled = bits & ((1u64 << 26) - 1);

    // Denormalize: scale back from [0, 2^26) to [0, 180) and [0, 360)
    let normalized_lat = lat_scaled as f64 * 180.0 / (1u64 << 26) as f64;
    let normalized_lon = lon_scaled as f64 * 360.0 / (1u64 << 26) as f64;

    // Convert back to original ranges
    let lat = normalized_lat - 90.0;
    let lon = normalized_lon - 180.0;

    Coordinate { lat, lon }
}

/// Calculate Haversine distance between two coordinates
fn haversine_distance(coord1: Coordinate, coord2: Coordinate, unit: DistanceUnit) -> f64 {
    let lat1 = coord1.lat.to_radians();
    let lon1 = coord1.lon.to_radians();
    let lat2 = coord2.lat.to_radians();
    let lon2 = coord2.lon.to_radians();

    let dlat = lat2 - lat1;
    let dlon = lon2 - lon1;

    let a = (dlat / 2.0).sin().powi(2) + lat1.cos() * lat2.cos() * (dlon / 2.0).sin().powi(2);
    let c = 2.0 * a.sqrt().asin();

    c * unit.radius()
}

/// Result type for georadius queries
pub type GeospatialRadiusResult = (Vec<u8>, Option<f64>, Option<Coordinate>);

/// Geospatial statistics
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GeospatialStats {
    pub total_keys: usize,
    pub total_locations: usize,
    pub geoadd_count: usize,
    pub geodist_count: usize,
    pub georadius_count: usize,
    pub geopos_count: usize,
    pub geohash_count: usize,
}

/// Geospatial store backed by Sorted Sets
pub struct GeospatialStore {
    sorted_set_store: Arc<SortedSetStore>,
    stats: Arc<RwLock<GeospatialStats>>,
}

impl GeospatialStore {
    /// Create a new geospatial store
    pub fn new(sorted_set_store: Arc<SortedSetStore>) -> Self {
        Self {
            sorted_set_store,
            stats: Arc::new(RwLock::new(GeospatialStats {
                total_keys: 0,
                total_locations: 0,
                geoadd_count: 0,
                geodist_count: 0,
                georadius_count: 0,
                geopos_count: 0,
                geohash_count: 0,
            })),
        }
    }

    /// GEOADD - Add one or more geospatial items (latitude, longitude, member)
    /// Returns the number of elements added (not updated)
    pub fn geoadd(
        &self,
        key: &str,
        locations: Vec<(f64, f64, Vec<u8>)>,
        nx: bool,
        xx: bool,
        ch: bool,
    ) -> Result<usize> {
        // Validate coordinates
        for (lat, lon, _) in &locations {
            Coordinate::new(*lat, *lon)?;
        }

        let locations_count = locations.len();
        let mut added = 0;

        // Check if key exists before adding (for total_keys tracking)
        let key_existed = self.sorted_set_store.zcard(key) > 0;

        for (lat, lon, member) in locations {
            let score = coordinate_to_score(lat, lon);

            let mut opts = super::sorted_set::ZAddOptions::default();
            opts.nx = nx;
            opts.xx = xx;
            opts.ch = ch;

            let (added_count, _changed_count) =
                self.sorted_set_store.zadd(key, member, score, &opts);
            added += added_count;
        }

        // Update statistics
        {
            let mut stats = self.stats.write();
            stats.geoadd_count += locations_count;
            if added > 0 {
                stats.total_locations += added;
            }
            // If key didn't exist before and we added at least one item, increment total_keys
            if !key_existed && added > 0 {
                stats.total_keys += 1;
            }
        }

        Ok(added)
    }

    /// GEODIST - Calculate distance between two members
    /// Returns distance in requested unit, or None if either member doesn't exist
    pub fn geodist(
        &self,
        key: &str,
        member1: &[u8],
        member2: &[u8],
        unit: DistanceUnit,
    ) -> Result<Option<f64>> {
        let coord1 = self.get_coordinate(key, member1)?;
        let coord2 = self.get_coordinate(key, member2)?;

        let coord1 = match coord1 {
            Some(c) => c,
            None => return Ok(None),
        };
        let coord2 = match coord2 {
            Some(c) => c,
            None => return Ok(None),
        };

        let distance = haversine_distance(coord1, coord2, unit);

        // Update statistics
        {
            let mut stats = self.stats.write();
            stats.geodist_count += 1;
        }

        Ok(Some(distance))
    }

    /// Get coordinate for a member
    fn get_coordinate(&self, key: &str, member: &[u8]) -> Result<Option<Coordinate>> {
        let score = self.sorted_set_store.zscore(key, member);
        Ok(score.map(score_to_coordinate))
    }

    /// GEOPOS - Get coordinates of one or more members
    /// Returns vector of Option<Coordinate> (None if member doesn't exist)
    pub fn geopos(&self, key: &str, members: &[Vec<u8>]) -> Result<Vec<Option<Coordinate>>> {
        let mut results = Vec::new();

        for member in members {
            let coord = self.get_coordinate(key, member)?;
            results.push(coord);
        }

        // Update statistics
        {
            let mut stats = self.stats.write();
            stats.geopos_count += members.len();
        }

        Ok(results)
    }

    /// GEOHASH - Get geohash string for one or more members
    /// Returns vector of Option<String> (None if member doesn't exist)
    pub fn geohash(&self, key: &str, members: &[Vec<u8>]) -> Result<Vec<Option<String>>> {
        let mut results = Vec::new();

        for member in members {
            let coord = self.get_coordinate(key, member)?;
            if let Some(coord) = coord {
                // Use geohash crate to encode
                let coord_obj = Coord {
                    x: coord.lon,
                    y: coord.lat,
                };
                // Redis uses 11-character geohash by default
                // encode returns Result, unwrap for now (coordinates are already validated)
                match encode(coord_obj, 11) {
                    Ok(hash) => results.push(Some(hash)),
                    Err(_) => results.push(None), // Should not happen with valid coordinates
                }
            } else {
                results.push(None);
            }
        }

        // Update statistics
        {
            let mut stats = self.stats.write();
            stats.geohash_count += members.len();
        }

        Ok(results)
    }

    /// GEORADIUS - Query members within radius of given coordinate
    /// Returns vector of (member, distance) pairs sorted by distance
    #[allow(clippy::too_many_arguments)]
    pub fn georadius(
        &self,
        key: &str,
        center_lat: f64,
        center_lon: f64,
        radius: f64,
        unit: DistanceUnit,
        with_dist: bool,
        with_coord: bool,
        count: Option<usize>,
        sort: Option<&str>,
    ) -> Result<Vec<GeospatialRadiusResult>> {
        let center = Coordinate::new(center_lat, center_lon)?;
        let radius_meters = match unit {
            DistanceUnit::Meters => radius,
            DistanceUnit::Kilometers => radius * 1000.0,
            DistanceUnit::Miles => radius * 1609.34,
            DistanceUnit::Feet => radius * 0.3048,
        };

        // Get all members with their scores
        let all_members = self.sorted_set_store.zrange(key, 0, -1, true);
        let mut results = Vec::new();

        // Debug: Check if we have members and verify coordinate encoding/decoding
        if all_members.is_empty() {
            return Ok(results);
        }

        for scored_member in all_members {
            let member_coord = score_to_coordinate(scored_member.score);
            let distance_meters = haversine_distance(center, member_coord, DistanceUnit::Meters);

            if distance_meters <= radius_meters {
                let distance = if with_dist {
                    Some(haversine_distance(center, member_coord, unit))
                } else {
                    None
                };
                let coord = if with_coord { Some(member_coord) } else { None };
                results.push((scored_member.member, distance, coord));
            }
        }

        // Sort by distance (ascending by default)
        if let Some(sort_str) = sort {
            match sort_str.to_uppercase().as_str() {
                "ASC" => {
                    results.sort_by(|a, b| {
                        let dist_a = a.1.unwrap_or(f64::INFINITY);
                        let dist_b = b.1.unwrap_or(f64::INFINITY);
                        dist_a
                            .partial_cmp(&dist_b)
                            .unwrap_or(std::cmp::Ordering::Equal)
                    });
                }
                "DESC" => {
                    results.sort_by(|a, b| {
                        let dist_a = a.1.unwrap_or(f64::INFINITY);
                        let dist_b = b.1.unwrap_or(f64::INFINITY);
                        dist_b
                            .partial_cmp(&dist_a)
                            .unwrap_or(std::cmp::Ordering::Equal)
                    });
                }
                _ => {} // No sort
            }
        } else {
            // Default: sort by distance ascending
            results.sort_by(|a, b| {
                let dist_a = a.1.unwrap_or(f64::INFINITY);
                let dist_b = b.1.unwrap_or(f64::INFINITY);
                dist_a
                    .partial_cmp(&dist_b)
                    .unwrap_or(std::cmp::Ordering::Equal)
            });
        }

        // Apply count limit
        if let Some(limit) = count {
            results.truncate(limit);
        }

        // Update statistics
        {
            let mut stats = self.stats.write();
            stats.georadius_count += 1;
        }

        Ok(results)
    }

    /// GEORADIUSBYMEMBER - Query members within radius of given member
    /// Returns vector of (member, distance) pairs sorted by distance
    #[allow(clippy::too_many_arguments)]
    pub fn georadiusbymember(
        &self,
        key: &str,
        member: &[u8],
        radius: f64,
        unit: DistanceUnit,
        with_dist: bool,
        with_coord: bool,
        count: Option<usize>,
        sort: Option<&str>,
    ) -> Result<Vec<GeospatialRadiusResult>> {
        // Get center coordinate from member
        let center_coord = self.get_coordinate(key, member)?;
        let center = match center_coord {
            Some(c) => c,
            None => {
                return Err(SynapError::KeyNotFound(format!(
                    "Member not found in geospatial key: {}",
                    String::from_utf8_lossy(member)
                )));
            }
        };

        // Delegate to georadius
        self.georadius(
            key, center.lat, center.lon, radius, unit, with_dist, with_coord, count, sort,
        )
    }

    /// Get statistics
    pub fn stats(&self) -> GeospatialStats {
        self.stats.read().clone()
    }
}

File: sdks/python/synap_sdk/modules/geospatial.py
"""Geospatial operations for Synap SDK."""

from __future__ import annotations

from typing import TYPE_CHECKING, Literal, TypedDict

if TYPE_CHECKING:
    from synap_sdk.client import SynapClient


DistanceUnit = Literal["m", "km", "mi", "ft"]


class Location(TypedDict):
    """Geospatial location coordinate."""

    lat: float
    lon: float
    member: str


class Coordinate(TypedDict):
    """Geographic coordinate."""

    lat: float
    lon: float


class GeoradiusResult(TypedDict, total=False):
    """Result from georadius query."""

    member: str
    distance: float | None
    coord: Coordinate | None


class GeospatialStats:
    """Geospatial statistics."""

    def __init__(self, data: dict[str, int]) -> None:
        """Initialize GeospatialStats from response data."""
        self.total_keys = data.get("total_keys", 0)
        self.total_locations = data.get("total_locations", 0)
        self.geoadd_count = data.get("geoadd_count", 0)
        self.geodist_count = data.get("geodist_count", 0)
        self.georadius_count = data.get("georadius_count", 0)
        self.geopos_count = data.get("geopos_count", 0)
        self.geohash_count = data.get("geohash_count", 0)


class GeospatialManager:
    """Manage Geospatial operations (Redis-compatible).

    Geospatial operations allow location-based queries using latitude/longitude.

    Example:
        >>> async with SynapClient(config) as client:
        ...     await client.geospatial.geoadd("cities", [
        ...         {"lat": 37.7749, "lon": -122.4194, "member": "San Francisco"}
        ...     ])
        ...     distance = await client.geospatial.geodist("cities", "SF", "NY", "km")
        ...     results = await client.geospatial.georadius("cities", 37.7749, -122.4194, 100, "km")
    """

    def __init__(self, client: SynapClient) -> None:
        """Initialize GeospatialManager.

        Args:
            client: The Synap client instance
        """
        self._client = client

    async def geoadd(
        self,
        key: str,
        locations: list[Location],
        nx: bool = False,
        xx: bool = False,
        ch: bool = False,
    ) -> int:
        """Add geospatial locations (GEOADD).

        Args:
            key: Geospatial key
            locations: Array of locations (lat, lon, member)
            nx: Only add new elements (don't update existing)
            xx: Only update existing elements (don't add new)
            ch: Return count of changed elements

        Returns:
            Number of elements added

        Raises:
            ValueError: If coordinates are out of valid range

        Example:
            >>> added = await geospatial.geoadd("cities", [
            ...     {"lat": 37.7749, "lon": -122.4194, "member": "San Francisco"},
            ...     {"lat": 40.7128, "lon": -74.0060, "member": "New York"}
            ... ])
        """
        # Validate coordinates
        for loc in locations:
            if not -90 <= loc["lat"] <= 90:
                raise ValueError(f"Latitude must be between -90 and 90, got: {loc['lat']}")
            if not -180 <= loc["lon"] <= 180:
                raise ValueError(f"Longitude must be between -180 and 180, got: {loc['lon']}")

        response = await self._client.send_command(
            "geospatial.geoadd",
            {
                "key": key,
                "locations": locations,
                "nx": nx,
                "xx": xx,
                "ch": ch,
            },
        )
        return response.get("added", 0)

    async def geodist(
        self,
        key: str,
        member1: str,
        member2: str,
        unit: DistanceUnit = "m",
    ) -> float | None:
        """Calculate distance between two members (GEODIST).

        Args:
            key: Geospatial key
            member1: First member
            member2: Second member
            unit: Distance unit (default: "m")

        Returns:
            Distance in specified unit, or None if either member doesn't exist

        Example:
            >>> distance = await geospatial.geodist("cities", "San Francisco", "New York", "km")
        """
        response = await self._client.send_command(
            "geospatial.geodist",
            {"key": key, "member1": member1, "member2": member2, "unit": unit},
        )
        return response.get("distance")

    async def georadius(
        self,
        key: str,
        center_lat: float,
        center_lon: float,
        radius: float,
        unit: DistanceUnit = "m",
        *,
        with_dist: bool = False,
        with_coord: bool = False,
        count: int | None = None,
        sort: Literal["ASC", "DESC"] | None = None,
    ) -> list[GeoradiusResult]:
        """Query members within radius (GEORADIUS).

        Args:
            key: Geospatial key
            center_lat: Center latitude
            center_lon: Center longitude
            radius: Radius
            unit: Distance unit (default: "m")
            with_dist: Include distance in results
            with_coord: Include coordinates in results
            count: Maximum number of results
            sort: Sort order ("ASC" or "DESC")

        Returns:
            List of matching members with optional distance and coordinates

        Raises:
            ValueError: If coordinates are out of valid range

        Example:
            >>> results = await geospatial.georadius("cities", 37.7749, -122.4194, 100, "km",
            ...                                      with_dist=True)
        """
        if not -90 <= center_lat <= 90:
            raise ValueError(f"Latitude must be between -90 and 90, got: {center_lat}")
        if not -180 <= center_lon <= 180:
            raise ValueError(f"Longitude must be between -180 and 180, got: {center_lon}")

        payload: dict[str, any] = {
            "key": key,
            "center_lat": center_lat,
            "center_lon": center_lon,
            "radius": radius,
            "unit": unit,
            "with_dist": with_dist,
            "with_coord": with_coord,
        }

        if count is not None:
            payload["count"] = count
        if sort is not None:
            payload["sort"] = sort

        response = await self._client.send_command("geospatial.georadius", payload)
        return response.get("results", [])

    async def georadiusbymember(
        self,
        key: str,
        member: str,
        radius: float,
        unit: DistanceUnit = "m",
        *,
        with_dist: bool = False,
        with_coord: bool = False,
        count: int | None = None,
        sort: Literal["ASC", "DESC"] | None = None,
    ) -> list[GeoradiusResult]:
        """Query members within radius of given member (GEORADIUSBYMEMBER).

        Args:
            key: Geospatial key
            member: Center member
            radius: Radius
            unit: Distance unit (default: "m")
            with_dist: Include distance in results
            with_coord: Include coordinates in results
            count: Maximum number of results
            sort: Sort order ("ASC" or "DESC")

        Returns:
            List of matching members with optional distance and coordinates

        Example:
            >>> results = await geospatial.georadiusbymember("cities", "San Francisco", 50, "km",
            ...                                              with_dist=True)
        """
        payload: dict[str, any] = {
            "key": key,
            "member": member,
            "radius": radius,
            "unit": unit,
            "with_dist": with_dist,
            "with_coord": with_coord,
        }

        if count is not None:
            payload["count"] = count
        if sort is not None:
            payload["sort"] = sort

        response = await self._client.send_command("geospatial.georadiusbymember", payload)
        return response.get("results", [])

    async def geopos(self, key: str, members: list[str]) -> list[Coordinate | None]:
        """Get coordinates of members (GEOPOS).

        Args:
            key: Geospatial key
            members: Array of member names

        Returns:
            List of coordinates (None if member doesn't exist)

        Example:
            >>> coords = await geospatial.geopos("cities", ["San Francisco", "New York"])
        """
        response = await self._client.send_command(
            "geospatial.geopos", {"key": key, "members": members}
        )
        return response.get("coordinates", [])

    async def geohash(self, key: str, members: list[str]) -> list[str | None]:
        """Get geohash strings for members (GEOHASH).

        Args:
            key: Geospatial key
            members: Array of member names

        Returns:
            List of geohash strings (None if member doesn't exist)

        Example:
            >>> geohashes = await geospatial.geohash("cities", ["San Francisco", "New York"])
        """
        response = await self._client.send_command(
            "geospatial.geohash", {"key": key, "members": members}
        )
        return response.get("geohashes", [])

    async def geosearch(
        self,
        key: str,
        *,
        from_member: str | None = None,
        from_lonlat: tuple[float, float] | None = None,
        by_radius: tuple[float, DistanceUnit] | None = None,
        by_box: tuple[float, float, DistanceUnit] | None = None,
        with_dist: bool = False,
        with_coord: bool = False,
        with_hash: bool = False,
        count: int | None = None,
        sort: Literal["ASC", "DESC"] | None = None,
    ) -> list[GeoradiusResult]:
        """Advanced geospatial search (GEOSEARCH).

        Args:
            key: Geospatial key
            from_member: Center member (mutually exclusive with from_lonlat)
            from_lonlat: Center coordinates as (lon, lat) tuple (mutually exclusive with from_member)
            by_radius: Search by radius as (radius, unit) tuple
            by_box: Search by bounding box as (width, height, unit) tuple
            with_dist: Include distance in results
            with_coord: Include coordinates in results
            with_hash: Include geohash in results (not yet implemented)
            count: Maximum number of results
            sort: Sort order ("ASC" or "DESC")

        Returns:
            List of matching members with optional distance and coordinates

        Example:
            >>> results = await geospatial.geosearch(
            ...     "cities",
            ...     from_member="San Francisco",
            ...     by_radius=(50, "km"),
            ...     with_dist=True
            ... )
        """
        if from_member is None and from_lonlat is None:
            raise ValueError("Either 'from_member' or 'from_lonlat' must be provided")
        if by_radius is None and by_box is None:
            raise ValueError("Either 'by_radius' or 'by_box' must be provided")

        payload: dict[str, any] = {
            "key": key,
            "with_dist": with_dist,
            "with_coord": with_coord,
            "with_hash": with_hash,
        }

        if from_member is not None:
            payload["from_member"] = from_member
        if from_lonlat is not None:
            payload["from_lonlat"] = list(from_lonlat)
        if by_radius is not None:
            payload["by_radius"] = [by_radius[0], by_radius[1]]
        if by_box is not None:
            payload["by_box"] = [by_box[0], by_box[1], by_box[2]]
        if count is not None:
            payload["count"] = count
        if sort is not None:
            payload["sort"] = sort

        response = await self._client.send_command("geospatial.geosearch", payload)
        return response.get("results", [])

    async def stats(self) -> GeospatialStats:
        """Retrieve geospatial statistics.

        Returns:
            Geospatial statistics

        Example:
            >>> stats = await geospatial.stats()
            >>> print(stats.total_keys)
        """
        response = await self._client.send_command("geospatial.stats", {})
        return GeospatialStats(response)

